### ISSUE TITLE
PodDisruptionBudget updates are forbidden

### ISSUE THREAD

**Issue by JorritSalverda** [kubernetes/kubernetes#45398]
_PodDisruptionBudget_ is immutable at this moment. It would be a nice improvement to allow changes to it, at the least to the _minAvailable_ field. Not sure whether this counts as a _bug_ or a _feature request_ :)

After creating a _PodDisruptionBudget_ with the following manifest.

```yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: elasticsearch-master
  labels:
    app: elasticsearch
    role: master
spec:
  selector:
    matchLabels:
      app: elasticsearch
      role: master
  minAvailable: 2
```

When changing the spec to let's say `minAvailable: 3` and re-running `kubectl apply` it returns the following error:

```
The PodDisruptionBudget "elasticsearch-master" is invalid: spec: Forbidden: updates to poddisruptionbudget spec are forbidden.
```

**Kubernetes version**:

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:33:11Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.6", GitCommit:"114f8911f9597be669a747ab72787e0bd74c9359", GitTreeState:"clean", BuildDate:"2017-03-28T13:36:31Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Google Container Engine 1.6.2 with cos image**:
- **OS** (e.g. from /etc/os-release):

```
BUILD_ID=9000.84.2
NAME="Container-Optimized OS"
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=56
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=56
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=cos
```

- **Kernel** (e.g. `uname -a`):

```
Linux gke-production-europe-we-auto-scaling-917da0af-zzzj 4.4.21+ #1 SMP Fri Feb 17 15:34:45 PST 2017 x86_64 Intel(
R) Xeon(R) CPU @ 2.50GHz GenuineIntel GNU/Linux
```

**What happened**:

Changing the value of `spec.minAvailable` in a _PodDisruptionBudget_ failed due to updates being forbidden.

**What you expected to happen**:

At least `spec.minAvailable` being allowed to change because it's often used in conjunction with a _StatefulSet_ which is likely to grow over time by incrementing the number of replicas, at which time you also want to grow _minAvailable_ in the _PodDisruptionBudget_.

**Comment by JorritSalverda** [comment]
@davidopp can you get the right SIGs tagged? Thx

**Comment by davidopp** [comment]
@kubernetes/sig-apps-feature-requests

**Comment by foxish** [comment]
@JorritSalverda For your particular use case (scaling with a statefulset), we're adding `maxUnavailable` (https://github.com/kubernetes/kubernetes/issues/34776).

**Comment by JorritSalverda** [comment]
@foxish that sounds more applicable. Will it allow both an integer as well as a percentage, like _minAvailable_ does?

**Comment by foxish** [comment]
Yes, it will have similar semantics to `minAvailable`.

**Comment by deitch** [comment]
Either way, will`PodDisruptionBudget` be updatable?

FYI, I get the same error even if the update changes nothing. How do I update my kubernetes cluster in a CD pipeline with a `PodDisruptionBudget` in it? Normally, I can do `kubectl apply -f somefile.yml`, and it intelligently handles services, deployments, statefulsets, etc. But if there is a disruption budget in there, boom.

**Comment by kargakis** [comment]
@foxish @davidopp is there any reason not to do this? Especially now that we are going to add MaxUnavailable, people may want to change to the new field.

**Comment by foxish** [comment]
I had a conversation with @mml about this. It appears that there was a rationale behind immutability of the spec initially when the plan included a time component as well (allowing N evictions over M hours). There isn't a reason that maxUnav/minAv shouldn't be mutable that I can think of.

**Comment by klynch** [comment]
Agreed, it is a bit frustrating to have to force replace this type when most other types do not require special treatment.

**Comment by h0tbird** [comment]
Our CI/CD pipeline is using `helm` and we hit this issue as well.
```
Error: UPGRADE FAILED: PodDisruptionBudget.policy "my-release" is invalid: spec: Forbidden: updates to poddisruptionbudget spec are forbidden.
```
I don't know how to avoid this problem other than removing the `PodDisruptionBudget` object.

**Comment by mikebryant** [comment]
In particular, even updates to the `.metadata.labels` appear to be blocked, and the error message is wrong, as I'm not trying to update the `spec`

```
$ kubectl label poddisruptionbudget kafka-budget mike=test
The PodDisruptionBudget "kafka-budget" is invalid: spec: Forbidden: updates to poddisruptionbudget spec are forbidden.
```

**Comment by rdsubhas** [comment]
Hello folks, can anyone pls describe how to use `PodDisruptionBudget` with helm? Since updates are forbidden, we are forced to manually re-create it all the time which kills source management for this one resource.

**Comment by dmfenton** [comment]
I agree, it would be really nice to be able to `apply` these files after creation for CD workflows.

**Comment by kow3ns** [comment]
You should use maxUnavailable and set that number to the maximum number of disruptions the applications will tolerate. I would argue that minAvaialble is a design bug as this value is predicated on the number of replicas for the Deployment, DaemonSet, StatefulSet, .... However, maxUnavialable (as a percentage or fixed value) should not vary with the scale of the Pod collection. If you can tolerate 
We could look at allowing mutation to metadata.

**Comment by rdsubhas** [comment]
@kow3ns thanks for your response. Do you know if there is a common standard across kubernetes resources on mutation? Because this is the first example where one resource does not work with helm, state management and continuous delivery. It would help a lot if kubernetes resources are helm-friendly, but that means a HUGE investment/barrier of aligning across resources on the same style or standard.

Any hopes of getting a mutation policy that is helm-friendly please?

**Comment by enisoc** [comment]
@kow3ns After reading this thread, I'm convinced we need an answer for mutating PDBs through `kubectl apply` (and thus Helm). For things like Pods, it's ok that you can't mutate them because we have things like Deployment that allow you to declaratively transition to a new Pod template. There's no such thing for PDB, so we are essentially preventing users from managing PDBs declaratively. They would have to imperatively delete the old one when they make any change.

Operationally you have a fair point that updating maxUnavailable should be uncommon, but making it immutable results in poor UX for the times when it's necessary (e.g. even if there's only one value that makes sense, maybe you used the wrong value by mistake and need to fix it). @foxish seems to think there's no semantic reason we can't make PDB mutable, but even if there were some reason that all changes needed to occur by deleting/recreating the PDB, then I'd argue we need to provide a "PDB Deployment" controller that contains a mutable PDB template and does the delete/recreate for you.

**Comment by davidopp** [comment]
FWIW this discussion has now forked into two parallel threads, here and #25321. I'd suggest coalescing them into one place.

**Comment by rowleyaj** [comment]
On the linked thread it seems like here was chosen as the place to continue discussion on this topic. Are there any further updates to this discussion elsewhere, or any issues implementing changes?

To add a further discussion point - is there any current reason that PDBs should not be mutable?

@foxish's comment:
>It appears that there was a rationale behind immutability of the spec initially when the plan included a time component as well (allowing N evictions over M hours). There isn't a reason that maxUnav/minAv shouldn't be mutable that I can think of.

Seems to imply that this was made immutable for a potential feature that does not yet exist. Whereas the current immutable implementation raises very real issues with `kubectl apply` and helm usage now as noted in this thread. I'm perhaps over simplifying but I believe these should be considered bugs in the implementation without a controller as @enisoc mentions. That said is it not better to make this a mutable resource rather than work to add another resource/controller. This would somewhat depend on if the other feature has been planned or dropped?

**Comment by mward29** [comment]
This is still open and its still causing us headaches through our CI pipeline

**Comment by cblecker** [comment]
Repinging @kubernetes/sig-apps-feature-requests

**Comment by bahar-p** [comment]
Any update on this issue? We can't still update charts with PDB

**Comment by DaveWHarvey** [comment]
While I agree this restriction is tedious for helm charts, the fact that helm does not detect that the change will fail is also a problem:  rather than rejecting the upgrade, helm installs the upgrade, only to have it fail, and then forces you to roll it back.      If there are going to be immutable fields, then there should be a way for helm to determine that those fields are immutable prior to actually starting the upgrade.   Ideally, a helm diff would even flag these.

**Comment by jracollins** [comment]
I have a solution that seems to be working well for me: 

Similar to the `{{- if .Release.IsInstall -}}` workaround linked before, but with the ability to change on subsequent upgrades.

If you leverage [helm hooks](https://github.com/helm/helm/blob/master/docs/charts_hooks.md) to create the Pod Disruption Budget, you can delete and recreate it on every helm install/upgrade.

Just add these annotations to the PDB like below in your chart: 

```yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: {{ template "app.name" . }}
  release: {{ .Release.Name }}
  annotations:
    helm.sh/hook: post-upgrade, post-install
    "helm.sh/hook-delete-policy": "before-hook-creation"
spec:
  minAvailable: {{ default (sub .Values.replicaCount 1 ) .Values.PodDisruptionBudget.minAvailable }}
```

**Comment by ManiacDC** [comment]
For those having this issue with helm, you can use the --force argument during an upgrade to force a deletion/re-creation of the resource if apply fails. It isn't specific to PDB's, though.

https://github.com/helm/helm/issues/2914

**Comment by seanjfellows** [comment]
I don't understand how it has been 2 years and this particular config still requires special treatment? `kubectl apply` crashes even when there's no change to the PodDisruptionBudget!

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/69867 allows this update, and will be included in 1.15

**Comment by seanjfellows** [comment]
Wow, awesome, thanks for the quick reply!
